ray-curator/ray_curator/stages/classifiers/base.py

```diff
@@ -0,0 +1,246 @@
+
+import cudf
+import pandas as pd
+import torch
+from crossfit import op
+from huggingface_hub import PyTorchModelHubMixin
+from torch import nn
+from transformers import AutoModel
+
+from ray_curator.stages.base import CompositeStage, ProcessingStage
+from ray_curator.stages.resources import Resources
+from ray_curator.tasks import DocumentBatch
+
+from .crossfit_wrappers import CrossFitLabelerWrapper, CrossFitPredictorWrapper, CrossFitTokenizerWrapper
+
+
+@dataclass(kw_only=True)
+class DistributedDataClassifier(ProcessingStage[DocumentBatch, DocumentBatch]):
+    """Abstract class for running multi-node multi-GPU data classification"""
+
+    labels: list[str]
+    filter_by: list[str]
+    model_batch_size: int
+    out_dim: int
+    pred_column: str | list[str]
+    max_chars: int
+    device_type: str
+    autocast: bool
+
+    @property
+    def name(self) -> str:
+        msg = "Subclasses must implement this method"
+        raise NotImplementedError(msg)
+
+    def inputs(self) -> tuple[list[str], list[str]]:
+        return ["data"], []
+
+    def outputs(self) -> tuple[list[str], list[str]]:
+        # TODO: Check for prob_column too, if there is one
+        return ["data"], [self.pred_column] if isinstance(self.pred_column, str) else self.pred_column
+
+    @property
+    def resources(self) -> Resources:
+        """Resource requirements for this stage."""
+        # TODO: Check this
+        return Resources(gpu_memory_gb=_get_suggest_memory_for_classifier())
+
+    def process(self, batch: DocumentBatch) -> DocumentBatch | None:
+        """Run classifier on documents and filter if desired."""
+
+        df = batch.to_pandas()
+        result_df = self._run_classifier(df)
+
+        if self.filter_by is not None:
+            result_df = self._filter_documents(result_df)
+
+        result_df = result_df.to_pandas()
+
+        if len(result_df) == 0:
+            print(f"All documents filtered out for batch {batch.task_id}")
+            return None
+
+        # Create output batch
+        return DocumentBatch(
+            task_id=f"{batch.task_id}_{self.name}",
+            dataset_name=batch.dataset_name,
+            data=result_df,
+        )
+
+    @abstractmethod
+    def _run_classifier(self, df: pd.DataFrame | cudf.DataFrame) -> pd.DataFrame | cudf.DataFrame:
+        pass
+
+    def _filter_documents(
+        self,
+        df: pd.DataFrame | cudf.DataFrame,
+    ) -> pd.DataFrame | cudf.DataFrame:
+        filter_by = self.filter_by
+        if isinstance(filter_by, str):
+            return df[df[self.pred_column].astype(str) == filter_by]
+        elif isinstance(filter_by, list):
+            return df[df[self.pred_column].isin(filter_by)]
+
+        msg = "filter_by must be a string or list type"
+        raise TypeError(msg)
+
+    def get_labels(self) -> list[str]:
+        return self.labels
+
+
+@dataclass(kw_only=True)
+class StreamingDataClassifier(CompositeStage[DocumentBatch, DocumentBatch]):
```

Comment:
Composite stage implementation. Playing around with just importing the `Tokenizer` step from CrossFit versus a CPU-only tokenizer step (see `Tokenizer` class in this file).